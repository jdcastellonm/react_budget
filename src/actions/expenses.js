// import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});
export const startAddExpense = ((expenseData = {}) => { // use thunk to return a function that will dispatch the action
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = {description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) => { // return the then() to allow promise chaining
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
});

export const removeExpense = (id) => ({
    type: 'REMOVE_EXPENSE',
    id
});
export const editExpense = (id, updates) => ({ // 'updates' is an object containing the updated properties
    type: 'EDIT_EXPENSE',
    id,
    updates
});