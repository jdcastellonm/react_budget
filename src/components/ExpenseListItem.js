import React from 'react';
import {Link} from 'react-router-dom';

// the spread can also be destructured like {description, amount, createdAt}
// instead of using props.*
const ExpenseListItem = (props) => (
    <div>
        <Link to={`/edit/${props.id}`}>
            <h3>{props.description}</h3>
        </Link>
        <p>{props.amount} -- {props.createdAt}</p>
    </div>
)
// export default connect()(ExpenseListItem); // access to state directly is not needed here, just the dispatch() (used when the remove button was here)
export default ExpenseListItem;