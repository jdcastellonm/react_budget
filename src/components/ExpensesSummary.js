import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectedExpenses from '../selectors/expenses';
import getExpenseTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({expenseCount, expensesTotal}) => { // destructuring props
    const word = expenseCount === 1 ? 'expense' : 'expenses';
    const totalFormatted = numeral(expensesTotal / 100).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {expenseCount} {word} totalling {totalFormatted}</h1>
        </div>
    )
};

const mapStateToProps = (state) => {
    const visibleExpenses = selectedExpenses(state.expenses, state.filters); // reminder this returns a filtered array
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpenseTotal(visibleExpenses) // get the sum from the filtered array
    }
};

export default connect(mapStateToProps)(ExpensesSummary);