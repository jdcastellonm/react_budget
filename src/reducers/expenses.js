
//expenses reducer. the default state is an empty array since it's an array of expense objects
export default (state = [], action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]; // concat returns a new array, it doesn't modify the original, but spread is better
        case 'REMOVE_EXPENSE':
            return state.filter((current) => {
                return current.id !== action.id;
            });
        case 'EDIT_EXPENSE':
            return state.map((current) => {
                if (action.id === current.id) {
                    return {...current, ...action.updates} // we use spread objects to create new ones and avoid mutating existing state
                } else {
                    return current;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};
