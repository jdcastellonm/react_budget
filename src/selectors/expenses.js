import moment from 'moment';

// filter expenses function. uses array filter method to get filter results
export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((current) => {
        const createdAtMoment = moment(current.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; 
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = current.description.toLowerCase().includes(text);

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        if (sortBy === 'amount') {
            return b.amount - a.amount;
        }
    })
}