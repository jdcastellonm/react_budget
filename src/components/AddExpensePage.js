import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

// refactor to class to use mapDispatchToProps and make testing easier
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                add expense component
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

// const AddExpensePage = (props) => (
//     <div>
//         add expense component
//         <ExpenseForm onSubmit={(expense) => { // 'expense' here comes from the form
//             props.dispatch(addExpense(expense));
//             props.history.push('/');
//         }}/>
//     </div>
// )

const mapDispatchToProps = (dispatch) => ({ // allows to call dispatch using simpler props.onSubmit syntax in the class, for easier testing
    addExpense: (expense) => dispatch(addExpense(expense))
});
export default connect(undefined, mapDispatchToProps)(AddExpensePage);