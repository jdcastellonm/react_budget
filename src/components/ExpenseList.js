import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (<p>No expenses.</p>) :
            (props.expenses.map((current) => {
                return <ExpenseListItem key={current.id} {...current}/>
            }))
        }
    </div>
);


// this is a hoc/decorator. the wrapper function gets returned from connect(), which is then immediately run to wrap
// the ExpenseList component. likely written this way to be able to make use of the state when preparing the wrapper
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters) // calling the sort by filters function here before mapping to props
    }
};
export default connect(mapStateToProps)(ExpenseList);