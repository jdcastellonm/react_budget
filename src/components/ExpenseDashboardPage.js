import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary'; // the default export is connected to the store so no need to pass in props

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
)
export default ExpenseDashboardPage;