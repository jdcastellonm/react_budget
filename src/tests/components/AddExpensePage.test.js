import React from 'react';
import {shallow} from 'enzyme';
import {AddExpensePage} from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

// globals for testing
// beforeEach runs before every test()
let addExpense, history, wrapper;
beforeEach(() => {
    addExpense = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history}/>);
})

test('render AddExpensePage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[2]);
})