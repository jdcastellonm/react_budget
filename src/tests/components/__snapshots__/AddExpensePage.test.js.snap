// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render AddExpensePage 1`] = `
<div>
  add expense component
  <ExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
